sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java

```diff
@@ -128,24 +131,36 @@ ContainersListBlobFlatSegmentResponse getBlobsInPageWithRestResponse(String page
                                                                          Integer maxResults,
                                                                          List<ListBlobsIncludeItem> include,
                                                                          Integer timeout,
-                                                                         String requestId) {
-        return this.getBlobsInPageWithRestResponseIntern(pageId, containerName,
+                                                                         String requestId,
+                                                                         CancellationToken cancellationToken) {
+        cancellationToken = cancellationToken == null ? CancellationToken.NONE : cancellationToken;
+        CallAndOptionalResult<ContainersListBlobFlatSegmentResponse> callAndOptionalResult =
+            this.getBlobsInPageWithRestResponseIntern(pageId, containerName,
             prefix,
             maxResults,
             include,
             timeout,
             requestId,
-            null).getResult();
+            null);
+
+        final Call call = callAndOptionalResult.getCall();
+        ((CancellationTokenImpl) cancellationToken).registerOnCancel(() -> {
+            call.cancel();
+        });
+
+        return callAndOptionalResult.getResult();
```

Comment:
I had not thought of this before: if the service call is going to be blocked and will either produce a result or fail before we reach this stage, is it any use registering the `onCancel` callback?

Reply:
In the case of blocking APIs like this, one way I see app using them is: 

```java
// UI Thread
CancellationToken token = CancellationToken.create();
```

```java
executorService.schedule(() -> {
    // Thread-Pool Thread
    client.getBlobProperties(.., cancellationToken);
});
```

```java
// UIThread
onButtonClient(() -> {
    cancellationToken.cancel();
});
```

Reply:
Wouldn't the call be completed before the `onCancel` callback with `call.cancel()` is registered on the token?

Reply:
I see you're right, in sync case the registration is delayed. I'm fixing it.